### PR TITLE
Reuse input related lists

### PR DESCRIPTION
**Pull request by UselessToucan**
As far as I see, current implementation allocates thousands of lists per second. I believe these allocations are unnecessary.

### CODE REVIEW

**Comment by bdach** [comment]
I can't say I'm entirely qualified for this, but this seems insanely dangerous for me. You've replaced locals with a field, and seeing as lists are returned by reference, you're therefore leaking a reference of the private list to the callers, who could do potentially *anything* with them as they wish outside of the locks within the class. If you were returning `IEnumerable` it would be still dangerous, but would require a cast, but this just feels like a big explosion waiting to happen.

Also, when submitting a PR with a performance fix I'd expect conclusive benchmarking results to prove the actual validity of it.

**Comment by smoogipoo** [comment]
As I mentioned in Discord, the way this is typically done is by using `List<T>.AsReadOnly()`. Benchmarking results would also be appreciated.

**Comment by bdach** [comment]
To add to the above however, usage of `AsReadOnly` would require a return type change, which therefore makes this a breaking API change requiring a transition period.

File: osu.Framework/Input/Handlers/InputHandler.cs [hunk 1]
```diff
@@ -21,20 +22,28 @@ public abstract class InputHandler : IDisposable
...
+        {
+            pendingInputsList = new List<IInput>();
+            readOnlyPendingInputsList = new ReadOnlyCollection<IInput>(pendingInputsList);
+        }
 
         /// <summary>
         /// Retrieve a list of all pending states since the last call to this method.
         /// </summary>
-        public virtual List<IInput> GetPendingInputs()
+        public virtual IList<IInput> GetPendingInputs()
```
Comment (bdach) [inline osu.Framework/Input/Handlers/InputHandler.cs]: While going from `List` to `IList` is a clever way to preserve backwards compatibility it feels like a half-measure to me. `IList` includes modifying members such as `Add()`, which then [throw `NotSupportedException`s](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1.system-collections-generic-icollection-t--add?view=netframework-4.8#System_Collections_ObjectModel_ReadOnlyCollection_1_System_Collections_Generic_ICollection_T__Add__0_) at call sites. Sure, people probably *shouldn't* be callling `Add()` on the return value of this, but I sincerely believe in making interfaces as safe as they possibly can be whenever possible.

**Comment by bdach** [comment]
Another consideration to test out: what happens if one of the callers happens to enumerate even the `ReadOnlyCollection` while updates of the underlying list are taking place in the critical section? I'm worried you'll get this:
```
InvalidOperationException: Collection was modified; enumeration operation may not execute
```

**Comment by UselessToucan** [comment]
Looks like these temporary lists can be removed at all
https://github.com/UselessToucan/osu-framework/commit/b0b5aba7d2de3c92e67965990ca3768fd970ad10

I played around with the visual tests manually and did not notice any problems. Also, there were no fails during test execution via NUnit.

**Comment by bdach** [comment]
The yield variant looks somewhat safer to me, although two notes:

* a benchmark would be nice to compare the cpu overhead of the underlying state machine execution vs the old plain lists,
* and I'm not a fan of the exclusive `pendingInputsRetrievalLock` lock going away. The yields should be under one to prevent possibly multiple concurrent consumers stealing each other's pending inputs.

**Comment by UselessToucan** [comment]
> * and I'm not a fan of the exclusive `pendingInputsRetrievalLock` lock going away. The yields should be under one to prevent possibly multiple concurrent consumers stealing each other's pending inputs.

I removed it because I was not sure that it is usable with `yield`. After [some reading](https://csharpindepth.com/Articles/IteratorBlockImplementation)("And finally...
" chapter) I found out that it actually can be used, but there is one problem:

> If you require some more code to be executed at some point after the value is yielded, you're in trouble: you can't guarantee it will happen.

I.e. it is not guaranteed that the lock will be released.

**Comment by iiSaLMaN** [comment]
Not a per-frame allocation but `MouseButtonEventManager` clones the positional input queue every mouse down handling (check `HandleMouseDown` and `HandleMouseUp`)

**Comment by bdach** [comment]
I can't stress it enough that in my opinion these sorts of changes *aren't valid* until the current implementation is *empirically* shown to incur a significant performance penalty compared to any suggested alternatives. None of the options proposed here to eliminate the allocations are as safe in usage as the original, and without benchmarking I can't even begin to guess if the trade-off is worth it. Yes, allocations every frame sounds bad - but if it turned out to take, say, even 5% of frame time, I'd probably take that over concurrency bugs from less safe solutions any day.

**Comment by peppy** [comment]
Agree. [Still waiting for benchmarks](https://github.com/ppy/osu-framework/pull/2891#issuecomment-541396972).

**Comment by UselessToucan** [comment]
I extracted `InputHandler.GetPendingInputs` method into a [benchmark project](https://github.com/UselessToucan/InputHandler.Benchmark).

Maybe I was mistaken somewhere, but here are the results I got.
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18362
AMD Ryzen 5 2600, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.100
  [Host]     : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT
  Job-VJOQWX : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT

InvocationCount=1  IterationCount=100  UnrollFactor=1  

```
|                        Method |     Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |---------:|---------:|---------:|------:|------:|------:|----------:|
|       IterateListInputHandler | 43.19 us | 2.006 us | 5.724 us |     - |     - |     - |   16640 B |
| IterateSharedListInputHandler | 35.02 us | 2.136 us | 6.232 us |     - |     - |     - |      40 B |
|      IterateYieldInputHandler | 21.64 us | 1.585 us | 4.624 us |     - |     - |     - |      56 B |

**Comment by peppy** [comment]
optimally you'd add a class to the existing framework benchmark project, rather than extracting code out.

**Comment by bdach** [comment]
To sum up:

* Arguments for this change:
  * Saves ~50% of execution time compared to current method
  * Seems to work fine after some in-the-wild testing by @UselessToucan (I haven't done any)
  * Less memory pressure due to fewer allocations
* Arguments against this change:
  * Execution time is in the microseconds anyway, so it shouldn't have meaningful impact on frame time
  * Allocations aren't huge either, for 1000 inputs (which is a lot anyway) they're way under the LOH lower limit of ~85kB
  * Due to locking mechanism used slow consumers can starve others if there are concurrent multiple ones
  * Additionally, after looking at a [rough analogue on sharplab](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEAsBuWsQGZlkQGFEBvWRCxABwCcBLANwEMAXAU0RraYBMB7AHYAbAJ6IAigFc20gDx0BLAHyIA+gEdp0xAF5EAtgHcJWtvMVKAFAEpMMStXrN2nbv2Fi+AIwBWbAMYsagDOIgL+AEp8fEF6BsbefoE2dg7klLjIEDjmyogA4mwsAKJCbAC2bIrBAGp0TACadGxCPDbpFGT2DpRCfP4A1oiWqqHhUTHWHT1dPXOIhgAWdGXDGqYAdAAqNCIAImyaMmyWfJJBzDSILRVVLNZT3fOUs8/zKBDIAOzXZZWKqTeFAAvtMHKCnogIcCgA) the state machine generated by `yield return` captures `this` as a closure, therefore accessing the underlying queue every time. Since adding to the concurrent queue is not under any locks (read/write or exclusive), in an extremely pathological scenario where events arrive faster while they are consumed this will deadlock due to the while loop never terminating. The current implementation doesn't have this problem, as the manager itself just dumps the entire queue on its own terms, returns it as a safe copy, and the consumer can do whatever they want with it without interfering with others.

As for my personal opinion I remain weakly against this change as I consider it to be less correct than the current implementation while not providing meaningful enough gains.

**Comment by UselessToucan** [comment]
>     * Due to locking mechanism used slow consumers can starve others if there are concurrent multiple ones

As far as I know, pending inputs are not consumed concurrently. The only consumer is an `InputManager`. Even if there are multiple `InputManager`s in the scene graph, their `Update` methods are executed in a single update thread(i.e. they are executing sequentially).

**Comment by smoogipoo** [comment]
Here's my thoughts on this:

I don't mind this kind of optimisation even though it seems like a very little amount, since we do want updates to be as lean as possible.
Our target is as close to 0-GC as possible, and 1000Hz devices do exist, so the memory constraints will be hit. Not only is 0-GC a concern, but overall memory fragmentation is also, which will cause problems on mobile and potentially on x86 devices should we choose in the future to support them in the future.

However with that being said, this is all irrelevant as the benchmark is flawed. There are two primary issues here:
1. This counts as a micro-benchmark, and if you'll notice it's invoking the test method once per run, for 100 runs. This is due to the `IterationSetup` method and you're running into the same issue I mentioned [here](https://github.com/ppy/osu-framework/pull/3167). Do _not_ use `IterationSetup` unless you're doing really expensive computations. The only way I've found around this is to put your initialisation within the code itself, and adding a baseline method with just the initialisation itself.
2. To provide a true representation of the cost of this, the line calling `GetPendingInputs()` should be called several times (maybe assume a 1000FPS target). The allocation of the state machine is likely to be worse than the allocation of inputs.

**Comment by UselessToucan** [comment]
@smoogipoo please check [this](https://github.com/UselessToucan/InputHandler.Benchmark/commit/785b80448c9130b08e181db5d73c662df6ec2787) commit.
The results looks more correct now.
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18362
AMD Ryzen 5 2600, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.100
  [Host]     : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT
  DefaultJob : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT


```
|                        Method |     Mean |    Error |   StdDev | Ratio |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |---------:|---------:|---------:|------:|--------:|------:|------:|----------:|
|       IterateListInputHandler | 79.42 us | 0.950 us | 0.889 us |  1.00 | 17.2119 |     - |     - |  70.31 KB |
| IterateSharedListInputHandler | 75.17 us | 0.822 us | 0.769 us |  0.95 |  9.5215 |     - |     - |  39.06 KB |
|      IterateYieldInputHandler | 83.54 us | 1.565 us | 1.464 us |  1.05 | 13.3057 |     - |     - |  54.69 KB |

In my opinion, we should reuse `List<T>` as it was initially done here.

**Comment by bdach** [comment]
To offer something constructive, here's an ugly (but I *think* correct and more optimised) idea for consideration - having a list pool in which each list is associated with each individual consumer (no guarantees it compiles, works or is performant):

```csharp
private ConcurrentDictionary<object, List<IInput>> listPool;

public virtual IEnumerable<IInput> GetPendingInputs(object consumer)
{
    listPool.GetOrAdd(consumer, () => new List<IInput>());
    // lock, add inputs to list, return list
}
```

Perhaps the `GetOrAdd` should be under the exclusive lock too; not too sure that's necessary. But then again I still don't think the gain is worth the hassle and the ensuing API ugliness in the above proposal, nor the bugs that might come when reusing the one list.

**Comment by smoogipoo** [comment]
I think this has since been resolved via other means. Let's close for now.

**Pull request closed without merging**